File: iosApp/iosApp/Views/Clubs/BookSearchField.swift
import SwiftUI
import Shared
import DesignSystem

/// Search-and-select field for choosing a book when creating a session. Debounced (400ms)
/// free-text search yields a dropdown of results rendered via `.popover` — SwiftUI's own
/// built-in mechanism for floating content anchored to a view, escaping the enclosing
/// hierarchy's clipping/layout proposal. Two earlier hand-rolled attempts (an in-hierarchy
/// `.overlay`, then a manual `UIWindow`) both broke in ways that were never confirmed root-
/// caused, since neither could be debugged live on the physical test device. `.popover` sidesteps
/// that whole class of bug by using Apple's own implementation instead of custom coordinate-space/
/// window-level code. `.presentationCompactAdaptation(.popover)` (iOS 16.4+; this app targets
/// 18.2+) keeps it a true popover on iPhone instead of the default compact-size-class fallback,
/// which is a full-screen sheet. Selecting a result collapses the field into a locked "selected
/// book" row, mirroring web's `BookSearchInput`.
struct BookSearchField: View {
    @Binding var query: String
    let results: [Shared.Book]
    let selectedBook: Shared.Book?
    let isSearching: Bool
    let isRegistering: Bool
    let error: String?
    let onSearch: (String) -> Void
    let onSelect: (Shared.Book) -> Void
    let onClear: () -> Void

    @State private var isPopoverPresented = false

    private var trimmedQuery: String { query.trimmingCharacters(in: .whitespaces) }

    private var shouldShowDropdown: Bool {
        !trimmedQuery.isEmpty && (!results.isEmpty || !isSearching)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            if let selectedBook {
                SelectedBookRow(book: selectedBook, onClear: onClear)
            } else {
                InputField(
                    label: "Book",
                    value: $query,
                    helperText: (isSearching || isRegistering) ? "Searching…" : nil
                )
                .popover(isPresented: $isPopoverPresented, arrowEdge: .bottom) {
                    dropdownContent
                        .frame(width: 340)
                        .presentationCompactAdaptation(.popover)
                }
                .onChange(of: shouldShowDropdown) { _, newValue in isPopoverPresented = newValue }
                .task(id: query) {
                    guard !trimmedQuery.isEmpty else { return }
                    try? await Task.sleep(nanoseconds: 400_000_000)
                    if !Task.isCancelled { onSearch(query) }
                }
            }

            if let error {
                Text(error)
                    .kluvsStyle(KluvsTheme.typography.caption)
                    .foregroundColor(KluvsTheme.colors.danger)
            }
        }
    }

    @ViewBuilder
    private var dropdownContent: some View {
        if !results.isEmpty {
            ScrollView {
                LazyVStack(spacing: 0) {
                    ForEach(Array(results.enumerated()), id: \.offset) { _, book in
                        BookResultRow(book: book, onTap: { onSelect(book) })
                    }
                }
                .padding(.top, 8)
            }
            .frame(maxHeight: 240)
            .background(KluvsTheme.colors.card)
            .clipShape(RoundedRectangle(cornerRadius: 8))
        } else {
            Text("No books found for \"\(query)\"")
                .kluvsStyle(KluvsTheme.typography.caption)
                .foregroundColor(KluvsTheme.colors.contentMuted)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(12)
                .background(KluvsTheme.colors.card)
                .clipShape(RoundedRectangle(cornerRadius: 8))
        }
    }
}

private struct BookResultRow: View {
    let book: Shared.Book
    let onTap: () -> Void

    var body: some View {
        HStack(spacing: 12) {
            BookThumbnail(imageUrl: book.imageUrl)
            VStack(alignment: .leading, spacing: 2) {
                Text(book.title)
                    .kluvsStyle(KluvsTheme.typography.body.medium)
                    .foregroundColor(KluvsTheme.colors.content)
                    .lineLimit(1)
                Text(book.author)
                    .kluvsStyle(KluvsTheme.typography.caption)
                    .foregroundColor(KluvsTheme.colors.contentMuted)
                    .lineLimit(1)
                if let year = book.year {
                    Text(String(year.intValue))
                        .kluvsStyle(KluvsTheme.typography.caption)
                        .foregroundColor(KluvsTheme.colors.contentMuted)
                }
            }
            Spacer()
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 8)
        .contentShape(Rectangle())
        .onTapGesture(perform: onTap)
    }
}

private struct SelectedBookRow: View {
    let book: Shared.Book
    let onClear: () -> Void

    var body: some View {
        HStack(spacing: 12) {
            BookThumbnail(imageUrl: book.imageUrl)
            VStack(alignment: .leading, spacing: 2) {
                Text(book.title)
                    .kluvsStyle(KluvsTheme.typography.body.medium)
                    .foregroundColor(KluvsTheme.colors.content)
                    .lineLimit(1)
                Text(book.author)
                    .kluvsStyle(KluvsTheme.typography.caption)
                    .foregroundColor(KluvsTheme.colors.contentMuted)
                    .lineLimit(1)
            }
            Spacer()
            Icon(type: .close, contentDescription: "Change book", tint: KluvsTheme.colors.contentMuted)
                .frame(width: 18, height: 18)
                .contentShape(Rectangle())
                .onTapGesture(perform: onClear)
        }
        .padding(12)
        .background(KluvsTheme.colors.card)
        .clipShape(RoundedRectangle(cornerRadius: 8))
    }
}

private struct BookThumbnail: View {
    let imageUrl: String?

    var body: some View {
        Group {
            if let imageUrl, let url = URL(string: imageUrl) {
                AsyncImage(url: url) { phase in
                    switch phase {
                    case .success(let image):
                        image.resizable().aspectRatio(contentMode: .fill)
                    default:
                        BookCoverPlaceholder()
                    }
                }
            } else {
                BookCoverPlaceholder()
            }
        }
        .frame(width: 40, height: 60)
        .clipShape(RoundedRectangle(cornerRadius: 4))
    }
}

#Preview {
    VStack(spacing: 24) {
        BookSearchField(
            query: .constant("hobbit"),
            results: [
                Book(id: "1", title: "The Hobbit", author: "J.R.R. Tolkien", edition: nil, year: 1937, isbn: nil, pageCount: nil, imageUrl: nil, externalGoogleId: nil),
                Book(id: "2", title: "The Hobbit: Illustrated Edition", author: "J.R.R. Tolkien", edition: nil, year: 1997, isbn: nil, pageCount: nil, imageUrl: nil, externalGoogleId: nil),
            ],
            selectedBook: nil,
            isSearching: false,
            isRegistering: false,
            error: nil,
            onSearch: { _ in },
            onSelect: { _ in },
            onClear: {}
        )
        BookSearchField(
            query: .constant("The Hobbit"),
            results: [],
            selectedBook: Book(id: "1", title: "The Hobbit", author: "J.R.R. Tolkien", edition: nil, year: 1937, isbn: nil, pageCount: nil, imageUrl: nil, externalGoogleId: nil),
            isSearching: false,
            isRegistering: false,
            error: nil,
            onSearch: { _ in },
            onSelect: { _ in },
            onClear: {}
        )
    }
    .padding()
}

File: composeApp/src/androidMain/kotlin/com/ivangarzab/kluvs/ui/clubs/BookSearchField.kt
package com.ivangarzab.kluvs.ui.clubs

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.layout.onGloballyPositioned
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.PreviewLightDark
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Popup
import androidx.compose.ui.window.PopupProperties
import coil3.compose.SubcomposeAsyncImage
import com.ivangarzab.kluvs.designsystem.components.bookcover.BookCoverPlaceholder
import com.ivangarzab.kluvs.designsystem.components.fields.InputField
import com.ivangarzab.kluvs.designsystem.components.icons.Icon
import com.ivangarzab.kluvs.designsystem.components.icons.IconType
import com.ivangarzab.kluvs.designsystem.theme.KluvsTheme
import com.ivangarzab.kluvs.model.Book
import kotlinx.coroutines.delay

private const val BOOK_SEARCH_DEBOUNCE_MS = 400L

/**
 * Search-and-select field for choosing a book when creating a session — the combobox flagged
 * as "separate, unbuilt" in [com.ivangarzab.kluvs.designsystem.components.fields.SearchField]'s
 * doc comment. Debounced free-text search yields a dropdown of results rendered as a floating
 * [Popup] anchored above the field (drawn in its own window layer, so it neither pushes the
 * rest of the bottom sheet around nor gets clipped by its scroll container, and opens away from
 * the keyboard instead of under it). Selecting a result collapses the field into a locked
 * "selected book" row, mirroring web's `BookSearchInput`. Uses [InputField] rather than
 * [com.ivangarzab.kluvs.designsystem.components.fields.SearchField] so it matches the height of
 * neighboring form fields (e.g. the due-date [PickerField]) instead of SearchField's compact
 * top-app-bar sizing.
 */
@Composable
fun BookSearchField(
    query: String,
    onQueryChange: (String) -> Unit,
    onSearch: (String) -> Unit,
    results: List<Book>,
    selectedBook: Book?,
    isSearching: Boolean,
    isRegistering: Boolean,
    error: String?,
    onSelect: (Book) -> Unit,
    onClear: () -> Unit,
    modifier: Modifier = Modifier,
) {
    LaunchedEffect(query, selectedBook) {
        if (selectedBook == null && query.isNotBlank()) {
            delay(BOOK_SEARCH_DEBOUNCE_MS)
            onSearch(query)
        }
    }

    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(8.dp)) {
        if (selectedBook != null) {
            SelectedBookRow(book = selectedBook, onClear = onClear)
        } else {
            val density = LocalDensity.current
            var fieldWidthPx by remember { mutableIntStateOf(0) }
            var dropdownHeightPx by remember { mutableIntStateOf(0) }
            val showDropdown = query.isNotBlank() && (results.isNotEmpty() || !isSearching)

            Box(
                modifier = Modifier.onGloballyPositioned {
                    fieldWidthPx = it.size.width
                },
            ) {
                InputField(
                    label = "Book",
                    value = query,
                    onValueChange = onQueryChange,
                    helperText = if (isSearching || isRegistering) "Searching…" else null,
                )

                // Floats above the field, in its own window layer — this is deliberately a
                // Popup, not an inline sibling composable, so opening it never reflows or gets
                // clipped by the enclosing bottom sheet's scroll container.
                if (showDropdown) {
                    Popup(
                        alignment = Alignment.TopStart,
                        offset = IntOffset(0, -(dropdownHeightPx + with(density) { 4.dp.roundToPx() })),
                        properties = PopupProperties(focusable = false),
                    ) {
                        Box(
                            modifier = Modifier
                                .width(with(density) { fieldWidthPx.toDp() })
                                .onGloballyPositioned { dropdownHeightPx = it.size.height },
                        ) {
                            when {
                                results.isNotEmpty() -> {
                                    LazyColumn(
                                        modifier = Modifier
                                            .fillMaxWidth()
                                            .heightIn(max = 240.dp)
                                            .background(KluvsTheme.colors.card, RoundedCornerShape(8.dp)),
                                    ) {
                                        items(results, key = { it.id.ifEmpty { it.externalGoogleId ?: it.title } }) { book ->
                                            BookResultRow(book = book, onClick = { onSelect(book) })
                                        }
                                    }
                                }
                                else -> {
                                    Text(
                                        text = "No books found for \"$query\"",
                                        style = KluvsTheme.typography.caption,
                                        color = KluvsTheme.colors.contentMuted,
                                        modifier = Modifier
                                            .fillMaxWidth()
                                            .background(KluvsTheme.colors.card, RoundedCornerShape(8.dp))
                                            .padding(12.dp),
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }

        if (error != null) {
            Text(
                text = error,
                style = KluvsTheme.typography.caption,
                color = KluvsTheme.colors.danger,
            )
        }
    }
}

@Composable
private fun BookResultRow(book: Book, onClick: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(horizontal = 12.dp, vertical = 8.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        BookThumbnail(imageUrl = book.imageUrl, contentDescription = book.title)
        Column {
            Text(
                text = book.title,
                style = KluvsTheme.typography.body.medium,
                color = KluvsTheme.colors.content,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            Text(
                text = book.author,
                style = KluvsTheme.typography.caption,
                color = KluvsTheme.colors.contentMuted,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            if (book.year != null) {
                Text(
                    text = book.year.toString(),
                    style = KluvsTheme.typography.caption,
                    color = KluvsTheme.colors.contentMuted,
                )
            }
        }
    }
}

@Composable
private fun SelectedBookRow(book: Book, onClear: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(KluvsTheme.colors.card, RoundedCornerShape(8.dp))
            .padding(12.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        BookThumbnail(imageUrl = book.imageUrl, contentDescription = book.title)
        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = book.title,
                style = KluvsTheme.typography.body.medium,
                color = KluvsTheme.colors.content,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            Text(
                text = book.author,
                style = KluvsTheme.typography.caption,
                color = KluvsTheme.colors.contentMuted,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
        }
        Icon(
            type = IconType.Close,
            contentDescription = "Change book",
            tint = KluvsTheme.colors.contentMuted,
            modifier = Modifier
                .size(18.dp)
                .clickable(onClick = onClear),
        )
    }
}

@Composable
private fun BookThumbnail(imageUrl: String?, contentDescription: String) {
    SubcomposeAsyncImage(
        model = imageUrl,
        contentDescription = contentDescription,
        modifier = Modifier
            .width(40.dp)
            .aspectRatio(2f / 3f)
            .clip(RoundedCornerShape(4.dp)),
        contentScale = ContentScale.Crop,
        loading = { BookCoverPlaceholder(modifier = Modifier.fillMaxWidth()) },
        error = { BookCoverPlaceholder(modifier = Modifier.fillMaxWidth()) },
    )
}

@PreviewLightDark
@Composable
private fun Preview_BookSearchField_Results() = KluvsTheme {
    Box(modifier = Modifier.background(KluvsTheme.colors.background).padding(16.dp)) {
        BookSearchField(
            query = "hobbit",
            onQueryChange = {},
            onSearch = {},
            results = listOf(
                Book(id = "1", title = "The Hobbit", author = "J.R.R. Tolkien", year = 1937, isbn = null),
                Book(id = "2", title = "The Hobbit: Illustrated Edition", author = "J.R.R. Tolkien", year = 1997, isbn = null),
            ),
            selectedBook = null,
            isSearching = false,
            isRegistering = false,
            error = null,
            onSelect = {},
            onClear = {},
        )
    }
}

@PreviewLightDark
@Composable
private fun Preview_BookSearchField_Selected() = KluvsTheme {
    Box(modifier = Modifier.background(KluvsTheme.colors.background).padding(16.dp)) {
        BookSearchField(
            query = "The Hobbit",
            onQueryChange = {},
            onSearch = {},
            results = emptyList(),
            selectedBook = Book(id = "1", title = "The Hobbit", author = "J.R.R. Tolkien", year = 1937, isbn = null),
            isSearching = false,
            isRegistering = false,
            error = null,
            onSelect = {},
            onClear = {},
        )
    }
}

@PreviewLightDark
@Composable
private fun Preview_BookSearchField_Empty() = KluvsTheme {
    Box(modifier = Modifier.background(KluvsTheme.colors.background).padding(16.dp)) {
        BookSearchField(
            query = "zzzznotabook",
            onQueryChange = {},
            onSearch = {},
            results = emptyList(),
            selectedBook = null,
            isSearching = false,
            isRegistering = false,
            error = null,
            onSelect = {},
            onClear = {},
        )
    }
}
